README.md
# Auction House DataBase
Auction House DataBase (AHDB for short), records DB history and allows offline queries, for classic and more.

This Open Source addon is mostly to capture the evolution of the WoW classic economy from empty to mature.

We have a unique chance to record that history and that's what this addon attempts to accomplish

As you might know the Blizzard APIs to query the AH from the web (like The Undermine Journal does for instance) won't exist at launch and for some undetermined time after launch, so let's together create and maintain that DataBase.  (DB uploader in the works)

## What does it do?

AHDB shows a (moveable) button to take the next step (target the auctioneer, start scan, save,...)

If you configured it to do so, each time you open the auction house (if available and unless you the hold shift key), AHDB takes a full snapshot of your auction house and record it in your saved variables under the realm, faction and timestamp

You will be able to later query that DB even when not at the AH

## More information

Get the binary release using [curseforge](https://www.curseforge.com/wow/addons/auction-house-database)/twitch
client or on wowinterface

The source of the addon resides on https://github.com/mooreatv/AuctionDB
(and the MoLib library at https://github.com/mooreatv/MoLib)

Releases detail/changes are on https://github.com/mooreatv/AuctionDB/releases

The data can be processed using https://github.com/mooreatv/AHDBapp

Note: AHDB is unrelated to TSM's internal module AuctionDB (though there is a basic integration and TSM can now use some of our awesome scan data)


ChangeLog.txt
ChangeLog for Auction House DataBase (AHDB) by MooreaTv (moorea@ymail.com)

Coming next:
- Your input/suggestions welcome !
- See open issues for ideas https://github.com/mooreatv/AuctionDB/issues

v1.01.04 October 14th 2019
- Add ProjectAzilroka to optional deps so it can manage our minimap icon.
- [wip] Pickup/Use latest MoLib scrolling frame (for now just `/run AuctionDB:TableDemo()`)

v1.01.03 Sept 30th 2019
- Fix rounding issue in MoLib's :round() and improved AH info calc
  fixes https://github.com/mooreatv/BetterVendorPrice/issues/6

v1.01.02 Sept 25th 2019
- Now (optionally) waits for seller info (you can turn it off in options for fastest scan)
  to consider the scan complete - most of the work as usual is in MoLib.

v1.00.01 Sept 11th 2019
- Stable enough for 1.0!
- Make the dependency on BVP optional because curseforge/twitch is broken
- Trying to flip around retail and classic tags and default to see if it helps
- Now use periodic yield to avoid (rare) DCs, from MoLib - should fix #23
- Config option to not show the periodic "you can scan" chat text - Fixes #29

v0.14.00 Sept 9th 2019
- Added support for LibDBIcon / SexyMap through MoLib changes (Fixes #20)
- Added support to disable all AHDB temporary keybindings (Fixes #21)
v0.12.03 Sept 9th 2019
- Adding option to hide the big button per request (Fixes #16)
v0.12.01,.02 Sept 5th 2019
- Re tag to get curse to pickup the dependency
v0.12.00 Sept 5th 2019
- Fix for #18: Raw data was fine but AH tooltip and TSM pricing info could miss the min for the first auction
  found if it had multiple items
v0.11.05 Aug 29th 2019
- Don't print warning about scan not doable at every AH open (Fixed #15) 
v0.11.04 Aug 28th 2019
- Fixed minimap button and big button drag to move forgetting location on reload (thx Bayart for the report!)
v0.11.03 Aug 28th 2019
- Fix bogus "Can't scan outside of cities..."
v0.11.02 Aug 28th 2019
- Software is hard... Really fixed #12, #13, #14 now. Should not show button unless rested and should
  not spam the chat periodically either
v0.11.01 Aug 28th 2019
- Opsa... was still starting with button shown irrespective of resting state (!) - sorry - fixed now!
v0.11.00 Aug 27th 2019
- No more pause/hang on classic at the end of the first scan!
- Unlike in BfA which doesn't mind despite 100x more items, for classic we needed to not restore 
  the item update event immediately, yet keep the UI functional even if you search after 
  and hit enter in Name editbox for instance (still haven't found how to hide previous results
  if you manual search first and then do a full scan but it's only cosmetically wrong in that case)
- Fix the inital /tar -> /scan button binding
- Detect "in combat" overlapping with start/end of resting to avoid addon blocked errors

v0.10.00 Aug 27th 2019
- Don't show the big pulsating button unless IsResting() (ie in city/inns) - Fixing Issue #12
- Don't bind key for full scan unless at AH
- Option to turn off the /tar auctioneer binding
- Now works even if another addon (like aux-addon) hides the standard UI (moved ah open/close events to MoLib)
- Also from MoLib: Bug fixes on normally shrinking auctions list 
  (in classic unlike BfA it seems the list refreshes faster despite having fewer auctions)

v0.09.02 Aug 26th 2019
- Pick up another MoLib small fix (round to .1 copper when dividing (avoids 16.6666666 copper), 
  coding error in 1 of the warnings)
v0.09.01 Aug 26th 2019
- Pick up MoLib fix with typo in classic code path that couldn't be tested until now (!) sorry!
v0.09.00 Aug 26th 2019
- Rebranded from short name "AuctionDB" to "AHDB" to avoid confusion with TSM's internal module of the same name
- Now include current bid in scan DB v6 and deserialization on load.
- (for now) Requires Better Vendor Price to display AH results information through tooltip
- Classic is today !

v0.08.00 Aug 25th 2019
- Pickup from MoLib: itemdb version 4 (20% shorter), and a bug fix for suffixid
- Better/more Localization strings

v0.07.01 Aug 24th 2019
(from MoLib)
- Fixed bogus realm mismatch warning
v0.07.00 Aug 23rd 2019
(from MoLib:)
- New ItemDB even more compact format 
- Scan data also better structure: group by item and seller, 
  now down to 1.5Mb a scan from 30Mb when I started.

v0.06.00 Aug 23rd 2019
- Pick up big AH changes in MoLib - https://github.com/mooreatv/MoLib/blob/master/ChangeLog.txt
- Also bind IWT (so hitting it 3 times starts a scan (target, open, start scan))
- New transparent, glowing button

AuctionDB.lua
--[[
   Auction House DataBase (AHDB) by MooreaTV moorea@ymail.com (c) 2019 All rights reserved
   Licensed under LGPLv3 - No Warranty
   (contact the author if you need a different license)

   Auction House DataBase, records DB history and offline queries, for classic and more

   Get this addon binary release using curse/twitch client or on wowinterface
   The source of the addon resides on https://github.com/mooreatv/AuctionDB
   (and the MoLib library at https://github.com/mooreatv/MoLib)

   Releases detail/changes are on https://github.com/mooreatv/AuctionDB/releases
   ]] --
--
-- our name, our empty default (and unused) anonymous ns
local addon, _ns = ...

-- Table and base functions created by MoLib
local ADB = _G[addon]
-- localization
ADB.L = ADB:GetLocalization()
local L = ADB.L

ADB.debugLocalization = false -- TODO: remove this line when dealing with localization

-- ADB.debug = 9 -- to debug before saved variables are loaded

ADB.slashCmdName = "ahdb"
ADB.addonHash = "c3f0f25"
ADB.savedVarName = "AuctionDBSaved"
ADB.name = "AHDB"
-- ADB.author = "MooreaTv" -- override default author

ADB.autoScan = false
ADB.autoScanDelay = 10
ADB.autoSave = false
ADB.showNewItems = 10 -- show first 10 new items seen
ADB.targetAuctioneer = true
ADB.showBigButton = true
ADB.disableKeybinds = false
ADB.showText = true

ADB.savePosSuffix = "buttonPos" -- button pos is button.name .. savePosSuffix

-- TODO: move most of this to MoLib

function ADB:SetupMenu()
  ADB:WipeFrame(ADB.mmb)
  ADB.minimapButtonAngle = 0
  local name = "AHDBminimapButton"
  local b = ADB:minimapButton(ADB[name .. ADB.savePosSuffix], name, "Interface/Addons/AuctionDB/AuctionDB.blp")
  b:SetScript("OnClick", function(_w, button, _down)
    if button == "RightButton" then
      ADB.Slash("config")
    else
      ADB:PrintDefault("AHDB: manual scan requested...")
      ADB.Slash("scan")
    end
  end)
  b.tooltipText =
    "|cFFF2D80CAuction House DataBase|r:\n" .. L["|cFF99E5FFLeft|r click to scan or open offline"] .. "\n" ..
      L["|cFF99E5FFRight|r click for options"] .. "\n\n" .. L["Drag to move this button."]
  b:SetScript("OnEnter", function()
    ADB:ShowToolTip(b, "ANCHOR_LEFT")
    ADB.inButton = true
  end)
  b:SetScript("OnLeave", function()
    GameTooltip:Hide()
    ADB.inButton = false
    ADB:Debug("Hide tool tip...")
  end)
  ADB:MakeMoveable(b, ADB.SavePositionCB) -- TODO move inside minimapButton to avoid isldbi nonsense
  ADB.mmb = b
end

function ADB.SavePositionCB(b, pos, _scale)
  ADB:SetSaved(b.name .. ADB.savePosSuffix, pos)
end

-- Events handling
ADB.doItButtonName = "AHDB_doItButton"

function ADB:DoItButton(cmd, msg, forceBind)
  local b = ADB.doItButton
  local ttip1 = "|cFFF2D80CAuction House DataBase|r: " ..
                  L["Action Button!\n\n|cFF99E5FFLeft|r click (or hit space, return or IWT key) to:"] .. "\n\n      "
  local ttip2 = "\n\n" .. L["Drag to move this button."]
  if not b then
    b = CreateFrame("Button", ADB.doItButtonName, UIParent, "InsecureActionButtonTemplate")
    ADB.doItButton = b
    b.name = "AHDBactionButton"
    -- local inset = CreateFrame("Frame", nil, b, "InsetFrameTemplate")
    -- inset:SetAllPoints()
    -- inset:SetIgnoreParentAlpha(true)
    -- b:SetFrameLevel(inset:GetFrameLevel() + 1)
    b:SetAttribute("type", "macro")
    b:SetSize(64, 64)
    local pos = ADB[b.name .. ADB.savePosSuffix]
    if pos then
      local pt, xOff, yOff = unpack(pos)
      b:SetPoint("TOPLEFT", nil, pt, xOff, yOff) -- dragging gives position from nil (screen)
    else
      b:SetPoint("TOP", 0, -10)
    end
    b:SetAlpha(.8)
    local t = b:CreateTexture(nil, "ARTWORK")
    t:SetTexture("Interface/Addons/AuctionDB/AuctionDB256.blp")
    t:SetAllPoints()
    local outline = b:CreateTexture(nil, "BACKGROUND")
    outline:SetTexture("Interface/Addons/AuctionDB/AuctionDB256glow.blp")
    outline:SetAllPoints()
    outline:SetAlpha(.1)
    local glow = b:CreateTexture(nil, "BACKGROUND")
    glow:SetTexture("Interface/Addons/AuctionDB/AuctionDB256glow.blp")
    -- glow:SetVertexColor(0.3,0.3,1) -- blueish glow
    glow:SetPoint("CENTER", 0, 0)
    glow:SetSize(68, 68)
    glow:SetBlendMode("ADD")
    glow:SetAlpha(0) -- start with no change
    glow:SetIgnoreParentAlpha(true)
    local ag = glow:CreateAnimationGroup()
    b.animationGroup = ag
    local anim = ag:CreateAnimation("Alpha")
    anim:SetFromAlpha(0)
    anim:SetToAlpha(0.4)
    ag:SetLooping("BOUNCE")
    anim:SetDuration(2)
    ag:Play()
    b:SetScript("OnEnter", function()
      ADB:ShowToolTip(b, "ANCHOR_RIGHT")
    end)
    b:SetScript("OnLeave", function()
      GameTooltip:Hide()
    end)
    b:SetScript("PostClick", function()
      ADB:HideDoItButton()
    end)
    ADB:MakeMoveable(b, ADB.SavePositionCB)
  end
  b.cmd = cmd
  b.tooltipText = ttip1 .. (msg or cmd) .. ttip2
  b:SetAttribute("macrotext", cmd)
  if (forceBind or self.ahShown) and not ADB.disableKeybinds then
    local iwtKey1 = GetBindingKey("INTERRACTTARGET")
    for _, key in next, {"ENTER", "SPACE", "RETURN", iwtKey1 or "."} do
      SetOverrideBindingClick(b, true, key, ADB.doItButtonName)
    end
    b.keyBound = true
  else
    if not ADB.inComnbat then
      ClearOverrideBindings(ADB.doItButton)
    end
    b.keyBound = false
  end
  if ADB.showBigButton then
    b:Show()
  else
    b:Hide()
  end
end

-- hides and most importantly clears temp key binds
function ADB:HideDoItButton()
  if ADB.doItButton then
    ADB:Debug("Hiding button")
    ClearOverrideBindings(ADB.doItButton)
    ADB.doItButton.keyBound = false
    ADB.doItButton:Hide()
  end
end

function ADB:ScanFrame()

end
-- Thresholds - TODO: add ui for thresholds
ADB.buyoutProfit = 48 -- 19 copper
ADB.bidProfit = 78 -- 69 copper

-- ADB.sendTo = "OFFICER"

function ADB:checkAuction(timeLeft, itemCount, minBid, buyoutPrice, bidAmount, minIncrement, ourBid, itemLink,
                          _auctionIndex)
  local _, _, _, _, _, _, _, _, _, _, vendorUnitPrice = GetItemInfo(itemLink)
  if not vendorUnitPrice or vendorUnitPrice <= 0 then
    ADB:Debug(4, "no vendor unit price (yet?) for % : %", itemLink, vendorUnitPrice)
    return -- not vendorable
  end
  if buyoutPrice > 0 then
    local vendorProfit = vendorUnitPrice * itemCount - buyoutPrice
    if vendorProfit > ADB.buyoutProfit then
      ADB:Debug("vendor buyout: % buyoutProfit=% vup=% buyout=% itemCount=% isOurBid=%", itemLink, vendorProfit,
                vendorUnitPrice, buyoutPrice, itemCount, ourBid)
      if not ourBid then
        ADB:PrintDefault("AHDB: |cFF00FF00Buyout|r Auction " .. itemLink .. "x% for " ..
                           GetCoinTextureString(buyoutPrice) .. " < vendor by " .. GetCoinTextureString(vendorProfit),
                         itemCount)
        -- Send to sendTo
        if ADB.sendTo then
          local msg = ADB:format("AHDB: Buyout Auction " .. itemLink .. "x% for " .. GetCoinText(buyoutPrice) ..
                                   " < vendor by " .. GetCoinText(vendorProfit), itemCount)
          SendChatMessage(msg, ADB.sendTo)
        end
      end
    end
  end
  if timeLeft ~= 1 then
    return -- don't look at bid for longer auctions
  end
  -- bid check, includes increment
  local bid = minBid
  if bidAmount > 0 then
    bid = bidAmount + minIncrement
  end
  local vendorProfit = vendorUnitPrice * itemCount - bid
  if vendorProfit > ADB.bidProfit then
    ADB:Debug("vendor bid: % bidProfit=% vup=% bid=% itemCount=% isOurBid=%", itemLink, vendorProfit, vendorUnitPrice,
              bid, itemCount, ourBid)
    if not ourBid then
      ADB:PrintDefault("AHDB: |cFFFF0000Short|r Auction " .. itemLink .. "x% bid " .. GetCoinTextureString(bid) ..
                         " < vendor by " .. GetCoinTextureString(vendorProfit), itemCount)
      -- Send to sendTo  -- GetCoinText
      if ADB.sendTo then
        local tmsg = ADB:format(
                       "AHDB: Short Auction " .. itemLink .. "x% bid " .. GetCoinText(bid) .. " < vendor by " ..
                         GetCoinText(vendorProfit), itemCount)
        SendChatMessage(tmsg, ADB.sendTo)
      end
    end
  end
end

-- original version
local auctionEntry = ADB.auctionEntry

-- vendor check "hook"
function ADB:auctionEntry(...)
  if ADB.devMode then
    -- this isn't ready yet (causes hangs with large auctions)
    ADB:checkAuction(...)
  end
  return auctionEntry(self, ...)
end

function ADB:Execute(cmd, msg, forceBind)
  -- don't respam
  if ADB.doItButton and ADB.doItButton:IsVisible() and ADB.doItButton.cmd == cmd and
    (ADB.doItButton.keyBound or not self.ahShown) then
    ADB:Debug("Same cmd " .. cmd .. " for button already visible and key bound % or not at ah %, ignoring",
              ADB.doItButton.keyBound, self.ahShown)
    return
  end
  local txt = cmd
  if msg then
    txt = msg .. " (" .. cmd .. ")"
  end
  local extra = ""
  if not forceBind and not self.ahShown then
    extra = L["When at the AH: "]
  end
  if ADB.showText then
    ADB:PrintDefault("AHDB: " .. extra .. L["click the button, or hit space or enter or IWT to "] .. txt)
  else
    ADB:Debug("Not showing text %", txt)
  end
  ADB:DoItButton(cmd, msg, forceBind)
end

function ADB:AfterSavedVars()
  if self.savedVar and self.savedVar.ah then
    ADB:AHRestoreData()
  end
end

local additionalEventHandlers = {

  PLAYER_ENTERING_WORLD = function(_self, ...)
    ADB:Debug("OnPlayerEnteringWorld " .. ADB:Dump(...))
    ADB:CreateOptionsPanel()
    ADB:SetupMenu()
    ADB.currentlyResting = IsResting()
    ADB:Debug("Initial resting is %", ADB.currentlyResting)
    if not ADB.currentlyResting then
      return
    end
    if ADB.targetAuctioneer then
      ADB:Execute("/tar " .. L["auctioneer"], L["Target the Auctioneer"], true) -- true == do bind even not at AH
    elseif ADB:AHfullScanPossible() then
      ADB:MaybeStartScan("enter world, resting")
    end
  end,

  DISPLAY_SIZE_CHANGED = function(_self)
    if ADB.mmb then
      ADB:SetupMenu() -- should be able to just RestorePosition() but...
    end
  end,

  UI_SCALE_CHANGED = function(_self, ...)
    ADB:DebugEvCall(1, ...)
    if ADB.mmb then
      ADB:SetupMenu() -- buffer with the one above?
    end
  end,

  PLAYER_REGEN_DISABLED = function(_self)
    ADB:Debug("Combat on")
    ADB:HideDoItButton()
    ADB.inCombat = true
  end,

  PLAYER_REGEN_ENABLED = function(_self)
    ADB:Debug("Combat off")
    ADB.inCombat = false
  end,

  PLAYER_UPDATE_RESTING = function(_self)
    local nowResting = IsResting()
    ADB:Debug("Change in resting (or initial) resting is % to %", ADB.currentlyResting, nowResting)
    if ADB.currentlyResting == nowResting then
      return -- no actual changes (happens in inns sometimes (!))
    end
    ADB.currentlyResting = nowResting
    if nowResting and not ADB.inCombat then
      if ADB:AHfullScanPossible() then
        ADB:MaybeStartScan("resting")
      end
    else
      ADB:HideDoItButton()
    end
  end
}

function ADB:AHOpenCB()
  ADB:Debug("AHDB AH open cb")
  ADB:MaybeStartScan("ah now open", true)
end

function ADB:AHCloseCB()
  ADB:PrintDefault("AHDB " .. L["AH closed"])
  ADB:HideDoItButton()
end

ADB:RegisterEventHandlers(additionalEventHandlers)

--
function ADB.Ticker() -- dot as it's ticker function
  ADB:Debug("Periodic ticker - scan possible: %", ADB:AHfullScanPossible())
  if ADB:AHfullScanPossible() and ADB.currentlyResting then
    ADB:MaybeStartScan("ticker")
  end
end

ADB.tickerInterval = 120 -- do not make this too frequent! 2 minutes is plenty for a 1 scan/15 mins allowed anyway
ADB.ticker = C_Timer.NewTicker(ADB.tickerInterval, ADB.Ticker)
--

function ADB:MaybeStartScan(msg, nowarning)
  msg = msg or ""
  self:Debug(2, "Called MaybeStartScan bcause " .. msg)
  if ADB.inCombat then
    ADB:Warning(L["Try again when not in combat..."])
    return
  end
  if not ADB.currentlyResting then
    ADB:Warning(L["Can't scan outside of cities..."])
    return
  end
  if not ADB:AHfullScanPossible() then
    if nowarning then
      ADB:Debug("can't do full scan, and no warning set - msg = " .. msg)
    else
      ADB:Warning(L["Can't do a full scan at this point, try later..."])
    end
    return
  end
  if not ADB.autoScan or not ADB.ahShown then
    ADB:Execute("/ahdb scan", L["Start a full scan"])
    return
  end
  if IsShiftKeyDown() then
    ADB:Warning(L["Shift key is down so we're not starting a scan."])
    ADB:Execute("/ahdb scan", L["Start a full manual scan now"])
    return
  end
  ADB:PrintDefault(L["Starting full scan (hold shift next time to prevent it or turn off auto scan)"])
  ADB:AHSaveAll()
end

function ADB:AHendOfScanCB()
  if ADB.autoSave then
    -- C_UI.Reload()
    ADB:Execute("/reload", L["Save the scan data to SavedVariables"], true)
  end
end

function ADB:Help(msg)
  ADB:PrintDefault("AHDB: " .. msg .. "\n" .. "/ahdb config -- open addon config\n" ..
                     "/ahdb scan -- manual full scan\n" .. "/ahdb bug -- report a bug\n" ..
                     "/ahdb debug on/off/level -- for debugging on at level or off.\n" ..
                     "/ahdb version -- shows addon version")
end

function ADB.Slash(arg) -- can't be a : because used directly as slash command
  ADB:Debug("Got slash cmd: %", arg)
  if #arg == 0 then
    ADB:Help("commands, you can use the first letter of each:")
    return
  end
  local cmd = string.lower(string.sub(arg, 1, 1))
  local posRest = string.find(arg, " ")
  local rest = ""
  if not (posRest == nil) then
    rest = string.sub(arg, posRest + 1)
  end
  if cmd == "b" then
    local subText = L["Please submit on discord or https://|cFF99E5FFbit.ly/ahbug|r or email"]
    ADB:PrintDefault(L["AHDB bug report open: "] .. subText)
    -- base molib will add version and date/timne
    ADB:BugReport(subText, "c3f0f25\n\n" .. L["Bug report from slash command"])
  elseif cmd == "v" then
    -- version
    ADB:PrintDefault("AHDB " .. ADB.manifestVersion .. " (c3f0f25) by MooreaTv (moorea@ymail.com)")
  elseif cmd == "s" then
    -- scan
    ADB:AHSaveAll()
  elseif ADB:StartsWith(arg, "context") then
    ADB:AHContext()
  elseif cmd == "c" then
    -- Show config panel
    -- InterfaceOptionsList_DisplayPanel(ADB.optionsPanel)
    InterfaceOptionsFrame:Show() -- onshow will clear the category if not already displayed
    InterfaceOptionsFrame_OpenToCategory(ADB.optionsPanel) -- gets our name selected
  elseif ADB:StartsWith(arg, "debug") then
    -- debug
    if rest == "on" then
      ADB:SetSaved("debug", 1)
    elseif rest == "off" then
      ADB:SetSaved("debug", nil)
    else
      ADB:SetSaved("debug", tonumber(rest))
    end
    ADB:PrintDefault("AHDB debug now %", ADB.debug)
  else
    ADB:Help('unknown command "' .. arg .. '", usage:')
  end
end

-- Run/set at load time:

-- Slash

SlashCmdList["AuctionDB_Slash_Command"] = ADB.Slash

SLASH_AuctionDB_Slash_Command1 = "/ahdb"

-- Options panel

function ADB:CreateOptionsPanel()
  if ADB.optionsPanel then
    ADB:Debug("Options Panel already setup")
    return
  end
  ADB:Debug("Creating Options Panel")

  local p = ADB:Frame("AHDB")
  ADB.optionsPanel = p
  p:addText(L["AHDB options"], "GameFontNormalLarge"):Place()
  p:addText(L["Auction House DataBase: records DB history, offline queries and more."]):Place()
  p:addText(L["These options let you control the behavior of AHDB"] .. " " .. ADB.manifestVersion ..
              " c3f0f25"):Place()

  local autoScan = p:addCheckBox(L["Auto Scan"],
                                 L["Automatically scan the AH whenever possible, unless the |cFF99E5FFShift|r key is held"])
                     :Place(4, 20)

  local scanDelay = p:addSlider(L["Auto scan delay"], L["How long to wait for cancellation before scan start"], 2, 10,
                                1, L["2 sec"], L["10 sec"]):PlaceRight(60, 4)
  scanDelay:DoDisable() -- not used/working yet

  local autoSave = p:addCheckBox(L["Auto Save/Reload"],
                                 L["Automatically prompts for /reload in order to save the DataBase at the end of the scan"])
                     :PlaceRight(60, -4)

  local waitForSellers = p:addCheckBox(L["Wait for Seller information"],
                                       L["Slower initial scan per session but more complete information with all sellers"])
                           :Place(4, 20)

  local doTarget = p:addCheckBox(L["Target Auctioneer at load time"],
                                 L["Automatically prompts for targetting the auctioneer at /reload or login time."])
                     :Place(4, 20)

  local showBigButton = p:addCheckBox(L["Show the big action button"],
                                      L["Shows, if checked, the big button prompting you to go do a scan; hides if unchecked."])
                          :Place(4, 20)

  local disableKeybinds = p:addCheckBox(L["Disable key bindings"],
                                        L["Disable the automatic temporary keybinding when a scan is possible."])
                            :PlaceRight(40)

  local showText = p:addCheckBox(L["Show text about scan possible and commands"],
                                 L["Shows or disable the text indicating a scan is possible" ..
                                   " and which command will be executed when clicking."]):Place(4, 20)

  local newItems = p:addSlider(L["Show new items"], L["Shows never seen before items found in scan up to these many"],
                               0, 100, 5, L["None"]):Place(16, 30) -- need more vspace

  p:addText(L["Development, troubleshooting and advanced options:"]):Place(40, 20)

  p:addButton(L["Bug Report"], L["Get Information to submit a bug."] .. "\n|cFF99E5FF/ahdb bug|r", "bug"):Place(4, 20)

  local allowLDBI = p:addCheckBox(L["Use SexyMap/LDBIcon if available"],
                                  L["When checked and if LibDBIcon is installed, use it for minimap icon, otherwise use our code."])
                      :Place(4, 20)

  p:addButton(L["Reset minimap button"], L["Resets the minimap button to back to initial default location"], function()
    ADB:SetSaved("AHDBminimapButtonbuttonPos", nil)
    ADB:SetupMenu()
  end):Place(4, 20)

  local debugLevel = p:addSlider(L["Debug level"], L["Sets the debug level"] .. "\n|cFF99E5FF/ahdb debug X|r", 0, 9, 1,
                                 "Off"):Place(16, 30)

  function p:refresh()
    ADB:Debug("Options Panel refresh!")
    if ADB.debug then
      -- expose errors
      xpcall(function()
        self:HandleRefresh()
      end, geterrorhandler())
    else
      -- normal behavior for interface option panel: errors swallowed by caller
      self:HandleRefresh()
    end
  end

  function p:HandleRefresh()
    p:Init()
    debugLevel:SetValue(ADB.debug or 0)
    autoScan:SetChecked(ADB.autoScan)
    scanDelay:SetValue(ADB.autoScanDelay)
    autoSave:SetChecked(ADB.autoSave)
    doTarget:SetChecked(ADB.targetAuctioneer)
    newItems:SetValue(ADB.showNewItems)
    showBigButton:SetChecked(ADB.showBigButton)
    allowLDBI:SetChecked(ADB.allowLDBI)
    disableKeybinds:SetChecked(ADB.disableKeybinds)
    showText:SetChecked(ADB.showText)
    waitForSellers:SetChecked(ADB.ahWaitForSellers)
  end

  function p:HandleOk()
    ADB:Debug(1, "ADB.optionsPanel.okay() internal")
    --    local changes = 0
    --    changes = changes + ADB:SetSaved("lineLength", lineLengthSlider:GetValue())
    --    if changes > 0 then
    --      ADB:PrintDefault("ADB: % change(s) made to grid config", changes)
    --    end
    local sliderVal = debugLevel:GetValue()
    if sliderVal == 0 then
      sliderVal = nil
      if ADB.debug then
        ADB:PrintDefault("AHDB: options setting debug level changed from % to OFF.", ADB.debug)
      end
    else
      if ADB.debug ~= sliderVal then
        ADB:PrintDefault("AHDB: options setting debug level changed from % to %.", ADB.debug, sliderVal)
      end
    end
    ADB:SetSaved("debug", sliderVal)
    ADB:SetSaved("autoScan", autoScan:GetChecked())
    ADB:SetSaved("autoScanDelay", scanDelay:GetValue())
    ADB:SetSaved("autoSave", autoSave:GetChecked())
    ADB:SetSaved("targetAuctioneer", doTarget:GetChecked())
    ADB:SetSaved("showNewItems", newItems:GetValue())
    ADB:SetSaved("disableKeybinds", disableKeybinds:GetChecked())
    ADB:SetSaved("showText", showText:GetChecked())
    ADB:SetSaved("ahWaitForSellers", waitForSellers:GetChecked())
    if ADB:SetSaved("allowLDBI", allowLDBI:GetChecked()) == 1 then
      ADB:SetupMenu()
    end
    local show = showBigButton:GetChecked()
    ADB:SetSaved("showBigButton", show)
    if show then
      if ADB.doItButton then
        ADB.doItButton:Show()
      end
    else
      ADB:HideDoItButton()
    end
  end

  function p:cancel()
    ADB:PrintDefault("AHDB: options screen cancelled, not making any changes.")
  end

  function p:okay()
    ADB:Debug(3, "ADB.optionsPanel.okay() wrapper")
    if ADB.debug then
      -- expose errors
      xpcall(function()
        self:HandleOk()
      end, geterrorhandler())
    else
      -- normal behavior for interface option panel: errors swallowed by caller
      self:HandleOk()
    end
  end
  -- Add the panel to the Interface Options
  InterfaceOptions_AddCategory(ADB.optionsPanel)
end

-- bindings / localization
_G.AUCTIONDB = "AHDB"
_G.BINDING_HEADER_ADB = L["Auction House DataBase addon key bindings"]
_G.BINDING_NAME_ADB_SCAN = L["AHDB Scan"] .. " |cFF99E5FF/ahdb scan|r"
_G.BINDING_NAME_ADB_OPEN = L["AHDB Open"] .. " |cFF99E5FF/ahdb open|r"

-- ADB.debug = 2
ADB:Debug("ahdb main file loaded")
